### ISSUE TITLE
Inconsistent Editor when editing songs with variable-bitrate-encoded MP3

### ISSUE THREAD

**Issue by BLuedtke** [UltraStar-Deluxe/USDX#749]
### Actual behaviour
When editing songs with the built-in editor, I noticed some weird behavior. On some songs, playing a whole line ("P") and playing single notes do not match. I.e., the timing/delay is different. There seems to be an offset; and when singing the song, it is different again.   

### Expected behaviour

The timing is consistent between playing a single note/syllable and the whole line.

### Steps to reproduce
Note: Using the latest release on Windows 10 (though I think I've seen this happen on linux too, can test if wanted).   
1. Encode an existing song with variable bitrate encoding, for example with ffmpeg (`ffmpeg -i <input>  -codec:a libmp3lame -q:a 1 <output>`)
2. Open the ultrastardx.exe, go to the song, and enter the editor
3. Play a few lines both as the whole line, and play as individual syllables/notes. At least on some lines, a timing mismatch should occur (but not necessarily on all of them).
4. Convert the mp3 to constant bitrate (or use the original if that already was CBR) `ffmpeg -i <input>  -codec:a libmp3lame -b:a 256k <output>`. Then try the editor again; the mismatch should not occur anymore.

### Details

Provide some additional information:

- USDX version: v2023.9.0
- Operating System + version: Windows 10 (don't know the build atm)

**Comment by barbeque-squared** [comment]
Yeah this also happens on Linux. It's not exclusive to mp3; anything that can be vbr like m4a, opus, ogg also has this problem. From what I remember it's less of an issue when skipping through songs while singing, because it will/should just sync everything else to the audio again. So if you want to go 5 seconds forward, it doesn't really matter if it's 4 or 6 instead. Obviously this is not really useful in the editor.

I doubt it will ever be fixed though. Maybe we can detect that it's vbr and pop up a warning if you open it in the editor, but that's about it unless the decoders for USDX have options for more precise seeking.

When I'm working on stuff I usually just convert it to some format, ie mp3 CBR for mp3's or flac for everything else, and then delete the converted file afterwards and change back the `#MP3` tag to the VBR file.

**Comment by bohning** [comment]
VBR has always been an issue. The simple yet effective solution is to simply avoid VBR and always use CBR. The Syncer ensures CBR for mp3 (if you choose to convert m4a to mp3), but if your karaoke software permits, you shoul use m4a anyhow. AFAIK, there is no VBR in m4a.

**Comment by s09bQ5** [comment]
M4A files probably contain an index that maps time to file offset. We could try to build an index on the fly.

**Comment by BLuedtke** [comment]
Thanks for all the infos, so it is a VBR issue then. Batch converting them is easy, I think the issue is for new users to realize that this is a problem (took me quite a while at least, didn't really find anything about this with my naive search). However, I also don't really see a good place/page in the Wiki this information could be added to.

**Comment by BLuedtke** [comment]
I wonder how difficult it is _to detect_ that a sound file is VBR-encoded. If it is easy, maybe upon opening the editor, it could display a popup warning about this? For MP3, the header of each frame [contains bitrate info](https://stackoverflow.com/questions/196418/how-can-i-programmatically-determine-whether-an-mp3-file-is-cbr-or-vbr-preferr), so it is technically possible to detect - but I'm not sure if it would be easy to implement. I have no idea about M4a.   
Also, idk if "system-triggered" popups are hard or generally avoided in this software, which I'd find completely understandable.

**Comment by s09bQ5** [comment]
You have to go through all frames until you find one that does not use the same bitrate as the previous ones. Then you know that you are dealing with VBR. But if you are doing that, you might as well create an index.

**Comment by s09bQ5** [comment]
Alternatively we could decompress the whole file to RAM and seek in there. Should be no problem on today's computers.
